File: 113-path-sum-ii/path-sum-ii.py
class Solution(object):
    def pathSum(self, root, targetSum):
        result = []

        def dfs(node, path, curSum):
            if not node:
                return

            path.append(node.val)
            curSum += node.val

            if not node.left and not node.right and curSum == targetSum:
                result.append(list(path))  # copy!

            dfs(node.left, path, curSum)
            dfs(node.right, path, curSum)

            path.pop()  # backtrack

        dfs(root, [], 0)
        return result

File: 113-path-sum-ii/path-sum-ii.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    vector<vector<int>> pathSum(TreeNode* root, int targetSum) {
        vector<vector<int>> result;
        vector<int>path;

        function<void(TreeNode* ,long)> dfs = [&](TreeNode* node, long rem){
            if (!node) return;

            path.push_back(node->val);
            rem += node->val;
            
            if (!node->left and !node->right and rem==targetSum) result.push_back(path);
            dfs(node->left,rem);
            dfs(node->right,rem);
            path.pop_back();

        };
        dfs(root,0);
        return result;
        
    }
};


